Fix cross-Node service access when AntreaProxy is disabled

tnqn opened the pull request:
When AntreaProxy is disabled, if the reply traffic of a connection that
has been processed by iptables/ipvs rules (of kube-proxy) is received
from the tunnel interface, its destination MAC would be rewritten twice
because it would have both gatewayCTMark and macRewriteMark set. The
latter rewriting would overwrite the former one and would cause the
packets to be delivered to the destination Pod directly without doing
reversed NAT in the host netns.

This patch fixes it by making the pipeline rewrite the destination MAC
as most once. It moves the gatewayCTMark related MAC rewriting flow to
l3ForwardingTable, to make L3 forwarding decision in same table
uniformly. It also simplifies the two gatewayCTMark related flows by
matching the direction of traffic which ensures the flow doesn't apply
to traffic from the gateway interface.

e2e tests for cross-Node service access are added to cover this scenario
with AntreaProxy disabled.

Fixes #2319

Need to backport to 0.13, 1.0, and 1.1.

tnqn commented:
/test-all

tnqn commented:
/test-all

tnqn commented:
/test-whole-conformance

jianjuns commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -1223,6 +1181,27 @@ func (c *client) l3FwdFlowToGateway(localGatewayIPs []net.IP, localGatewayMAC ne
+	//  kube-proxy is used). In this case the destination IP address of the reply traffic is the Pod which initiated the
+	//  connection to the Service (no SNAT). We need to make sure that these packets are sent back through the gateway
+	//  so that the source IP can be rewritten (Service backend IP -> Service ClusterIP).
+	//  2) when hair-pinning is involved, i.e. for connections between 2 local Pods belonging to this Node and for which
```
"for connections between 2 local Pods belonging to this Node and for which NAT Bis performed"

How about: "connections between 2 local Pods on the Node, for which NAT is performed"?

tnqn replied:
Thanks, will update after running all CI tests

tnqn replied:
I found the first sentence is almost same as the one in ovs-pipeline, unified them. Not sure if you mean removing the leading "for" or not. The latest version is "when hair-pinning is involved, i.e. for connections between 2 local Pods, for which NAT is performed." Let me know if it works for you.

jianjuns replied:
I think we can remove the leading "for", but we could keep it too.

jianjuns reviewed (commented):
In the commit description:

>  It moves the gatewayCTMark related mac rewritten flow to
l3ForwardingTable

"mac rewritten flow" -> "MAC rewriting flow" or "MAC rewrite flow".

jianjuns commented:
@tnqn : question - is it a regression (hope it was not introduced by my SNAT changes..)? Otherwise, we did not find it earlier.

tnqn commented:
> @tnqn : question - is it a regression (hope it was not introduced by my SNAT changes..)? Otherwise, we did not find it earlier.

It was a regression introduced since 0.13.0. I described detail and why it was not caught by CI tests in https://github.com/antrea-io/antrea/issues/2319. I guess it's related to https://github.com/antrea-io/antrea/commit/53e407e2baee73cb29f2f804d7f7d11cb5421178.

tnqn commented:
/test-all

tnqn commented:
/test-ipv6-all

tnqn commented:
/test-all-ipv6

jianjuns reviewed (dismissed):
Would you revise "gatewayCTMark related mac rewritten flow" in the commit message to "gatewayCTMark related MAC rewrite flow" before merging?

tnqn commented:
> Would you revise "gatewayCTMark related mac rewritten flow" in the commit message to "gatewayCTMark related MAC rewrite flow" before merging?

Removed leading "for" and corrected the commit message, thanks for reminding.

tnqn commented:
/test-all

tnqn commented:
/test-e2e

tnqn commented:
/skip-e2e

e2e once succeeded and the latest failure was for "TestFlowAggregator" which has been tracked in https://github.com/antrea-io/antrea/issues/2283
